src/block/utils/BlockDisplacer.php

```diff
@@ -0,0 +1,40 @@
+<?php
+
+/*
+ *
+ *  ____            _        _   __  __ _                  __  __ ____
+ * |  _ \ ___   ___| | _____| |_|  \/  (_)_ __   ___      |  \/  |  _ \
+ * | |_) / _ \ / __| |/ / _ \ __| |\/| | | '_ \ / _ \_____| |\/| | |_) |
+ * |  __/ (_) | (__|   <  __/ |_| |  | | | | | |  __/_____| |  | |  __/
+ * |_|   \___/ \___|_|\_\___|\__|_|  |_|_|_| |_|\___|     |_|  |_|_|
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * @author PocketMine Team
+ * @link http://www.pocketmine.net/
+ *
+ *
+ */
+
+declare(strict_types=1);
+
+namespace pocketmine\block\utils;
+
+use pocketmine\block\Block;
+
+/**
+ * Indicates that a block can contain another block inside.
+ * Used in things like waterlogging & snowlogging.
+ */
+interface BlockDisplacer{
+
+	/**
+	 * @internal
+	 * Called when a block is loaded to update its contained block.
+	 * Do NOT use this in plugins, use your block provided API instead.
+	 */
+	public function setDisplacedBlock(?Block $block) : void;
```

Comment:
I can't think of a reason why we'd want a generalised version of this, unless it was also checking the acceptance conditions. Since it should anyway require checking the type of displacer to know whether the displaced block can coexist with it.